litebird_sim/mbs/mbs.py

```diff
@@ -0,0 +1,498 @@
+                    elif 'npy' in custom_instrumet:
+                        self.instrument = np.load(custom_instrument, allow_picke=True).item()
+                    else:
+                        raise NameError('Wrong instrument dictonary format')
+                else:
+                    self.instrument = {}
+                    if self.telescopes:
+                        channels = []
+                        for tel in self.telescopes:
+                            channels.append(self.imo.query(
+                                f'/releases/v{self.imo_version}/satellite/{tel}/instrument_info').metadata['channel_names'])
+                        channels = [item for sublist in channels for item in sublist]
+                    else:
+                        channels = self.channels
+                    for ch in channels:
+                        if 'L' in ch:
+                            tel = 'LFT'
+                        elif 'M' in ch:
+                            tel = 'MFT'
+                        elif 'H' in ch:
+                            tel = 'HFT'
+                        data_file = self.imo.query(f'/releases/v{self.imo_version}/satellite/{tel}/{ch}/channel_info')
+                        freq = data_file.metadata['bandcenter_ghz']
+                        freq_band = data_file.metadata['bandwidth_ghz']
+                        fwhm_arcmin = data_file.metadata['fwhm_arcmin']
+                        P_sens = data_file.metadata['pol_sensitivity_channel_ukarcmin']
+                        self.instrument[ch] = {'freq':freq, 'freq_band': freq_band, 'beam': fwhm_arcmin, 'P_sens': P_sens }
+
+    def check_parameters(self):
+        config_mbs = self.parameters
+        self.nside = config_mbs['general']['nside']
+        try: self.save = config_mbs['general']['save']
+        except KeyError:
+            self.save = True
+            log.info("setting save = True")
+        try: self.gaussian_smooth = config_mbs['general']['gaussian_smooth']
+        except KeyError:
+            self.gaussian_smooth = False
+            log.info("setting gaussian_smooth = False")
+        try: self.bandpass_int = config_mbs['general']['bandpass_int']
+        except KeyError:
+            self.bandpass_int = False
+            log.info("setting bandpass_int = False")
+        try: self.bandpass_int = config_mbs['general']['bandpass_int']
+        except KeyError:
+            self.bandpass_int = False
+            log.info("setting bandpass_int = False")
+        try: self.coadd = config_mbs['general']['coadd']
+        except KeyError:
+            if self.save==True:
+                self.coadd = False
+                log.info("setting coadd = False")
+            else:
+                self.coadd = True
+        try: self.parallel_mc = config_mbs['general']['parallel_mc']
+        except KeyError:
+            self.parallel_mc = False
+            log.info("setting parallel_mc = False")
```

Comment:
The code above can be made much simpler by using a dedicated dataclass:

```python
from dataclasses import dataclass, fields
from typing import Union

# Define this *outside* the class `Mbs`
@dataclass
class MbsParameters:
    nside: int
    save: bool = True
    gaussian_smooth: bool = False
    bandpass_int: bool = False
    coadd: bool
    parallel_mc: bool = False
    make_noise: bool = False
    nmc_noise: int = 1
    seed_noise: Union[int, None] = None
    N_split: Union[int, bool] = False
    # etc.

    def __post_init__(self): # This is some magic enabled by the "dataclasses" method
        # Here you can manually fix parameters after they have been initialized
        # by a call to the constructor or to the `from_dict` static method
       if self.N_split == 1:
           self.N_split = False

       # Etc.

    @staticmethod
    def from_dict(dictionary):
        result = MbsParameters()
        for param in fields(result):
            if param.name in dictionary:
                setattr(result, param.name, dictionary[param.name])

        # This is not a constructor, so we need to call it manually
        result.__post_init__()
        return result
```

This is the same approach used in the class `DetectorInfo` (file `litebird_sim/detectors.py`), and it has the following advantages:

1.   All the parameters used by the simulation are listed in one place, and their default values are stated explicitly and collected within a narrow piece of code.

2.   The user can get a list of the parameters accepted by the module, as well as their default values, writing

          ?MbsParameters

      at the IPython prompt.

2.   There is no need to manually read them one by one from a dictionary, as the `fields` method permits to iterate over them; in fact, the method `check_parameters` becomes:

      ```python
      def check_parameters(self):
          self.config = MbsParameters(self.parameters)
          # That's it! You can refer to any parameter later as `self.config.units`,
          # `self.config.parallel_mc`, etc.
      ```

      In fact, the method `check_parameters` becomes so trivial that it could probably be moved in the `Mbs.__init__` method.

3.   An object `MbsParameters` can be easily created on-the-fly, and this is probably useful in the unit tests for the class.